Git Beta: Zed crashes on viewing git removed multiple lines while column git blame is open

### Summary

<!-- Please insert a one line summary of the issue below -->

SUMMARY_SENTENCE_HERE

<!-- Be verbose: Include all steps necessary to reproduce from a clean Zed installation. -->
<!-- Code snippets are better than images, a repository link that reproduces the issue is ideal. -->

Steps to trigger the problem:
1. open zed
2. remove some lines of code (this also can be done while git blame column opened)
3. enable column git blame
4. now click on the red triangle that shows removed lines
5. zed crashes and closes 

Actual Behavior:
zed crashes when trying to view removed lines while column git blame is open

Expected Behavior:
zed should be able to show removed lines without problem

https://github.com/user-attachments/assets/56e0be63-1ac9-4e4d-8d72-0faa04ad12fe

**note zed works fine with new line addition and line modification**
**note it crashes if more than one line is removed** 
**note it only crashes if column git blame is open** 

<!--
  Is there anything additional necessary to reproduce this issue?
    - settings.json, keymap.json, .editorconfig etc?
    - Does it happen intermittently or only with specific projects / file types?
    - Have you found a workaround?

  Did you check your Zed.log to see if there is any relevant details there?
    - When including large items (videos, screenshots, logs, configs) please wrap with:

  <details><summary>See inside for XXXXYYY</summary>

  ```shell
  code
  ```

  </details>
-->


### Zed Version and System Specs

Zed: v0.176.2 (Zed) 
OS: Linux Wayland fedora 41
Memory: 15.2 GiB
Architecture: x86_64
GPU: Intel(R) UHD Graphics (CML GT2) || Intel open-source Mesa driver || Mesa 24.3.4

Thanks for the report, I can reproduce this on nightly 89d89b8b2d:

```
2025-03-05T09:56:17.62901-05:00 [ERROR] {
  "thread": "main",
  "payload": "cannot seek backward",
  "location_data": {
    "file": "crates/editor/src/git/blame.rs",
    "line": 201
  },
  "backtrace": [
    "backtrace::capture::Backtrace::create::h21b339776477734e+11679748",
    "backtrace::capture::Backtrace::new::hc8a089174c4cda25+11679556",
    "zed::reliability::init_panic_hook::{{closure}}::ha42663a5be87f8f1+88412376",
    "std::panicking::rust_panic_with_hook::hc9e2ec2a12b10e12+69496416",
    "std::panicking::begin_panic_handler::{{closure}}::h3b6449e2eb83bf4c+69495204",
    "std::sys::backtrace::__rust_end_short_backtrace::h38d85099ddd107a3+69489612",
    "_rust_begin_unwind+69494416",
    "core::panicking::panic_fmt::hd4003258f6ab7b29+98588236",
    "sum_tree::cursor::Cursor<T,D>::seek_internal::h655646cc80eb31b9+23079740",
    "editor::git::blame::GitBlame::blame_for_rows::{{closure}}::hf6f489e88ec8cf1c+24058444",
    "<editor::element::EditorElement as gpui::element::Element>::prepaint::{{closure}}::{{closure}}::{{closure}}::hbac4e3516102848f+23759284",
    "<editor::element::EditorElement as gpui::element::Element>::prepaint::{{closure}}::hd6eef2f4166280a7+23695300",
    "<editor::element::EditorElement as gpui::element::Element>::prepaint::he4556d523e99ac64+23694144",
    "<gpui::element::Drawable<E> as gpui::element::ElementObject>::prepaint::hcda246cd0acc2c82+22762184",
    "gpui::element::AnyElement::prepaint::h3dafbe732f7ac74a+35357644",
  ]
}
```

--- a/crates/collab/src/tests/editor_tests.rs
+++ b/crates/collab/src/tests/editor_tests.rs
@@ -2027,6 +2027,15 @@ async fn test_git_blame_is_forwarded(cx_a: &mut TestAppContext, cx_b: &mut TestA
         .unwrap()
         .downcast::<Editor>()
         .unwrap();
+    let buffer_id_b = editor_b.update(cx_b, |editor_b, cx| {
+        editor_b
+            .buffer()
+            .read(cx)
+            .as_singleton()
+            .unwrap()
+            .read(cx)
+            .remote_id()
+    });
 
     // client_b now requests git blame for the open buffer
     editor_b.update_in(cx_b, |editor_b, window, cx| {
@@ -2045,6 +2054,7 @@ async fn test_git_blame_is_forwarded(cx_a: &mut TestAppContext, cx_b: &mut TestA
                     &(0..4)
                         .map(|row| RowInfo {
                             buffer_row: Some(row),
+                            buffer_id: Some(buffer_id_b),
                             ..Default::default()
                         })
                         .collect::<Vec<_>>(),
@@ -2092,6 +2102,7 @@ async fn test_git_blame_is_forwarded(cx_a: &mut TestAppContext, cx_b: &mut TestA
                     &(0..4)
                         .map(|row| RowInfo {
                             buffer_row: Some(row),
+                            buffer_id: Some(buffer_id_b),
                             ..Default::default()
                         })
                         .collect::<Vec<_>>(),
@@ -2127,6 +2138,7 @@ async fn test_git_blame_is_forwarded(cx_a: &mut TestAppContext, cx_b: &mut TestA
                     &(0..4)
                         .map(|row| RowInfo {
                             buffer_row: Some(row),
+                            buffer_id: Some(buffer_id_b),
                             ..Default::default()
                         })
                         .collect::<Vec<_>>(),

--- a/crates/editor/src/git/blame.rs
+++ b/crates/editor/src/git/blame.rs
@@ -538,6 +538,7 @@ mod tests {
     use serde_json::json;
     use settings::SettingsStore;
     use std::{cmp, env, ops::Range, path::Path};
+    use text::BufferId;
     use unindent::Unindent as _;
     use util::{path, RandomCharIter};
 
@@ -555,16 +556,18 @@ mod tests {
     #[track_caller]
     fn assert_blame_rows(
         blame: &mut GitBlame,
+        buffer_id: BufferId,
         rows: Range<u32>,
         expected: Vec<Option<BlameEntry>>,
         cx: &mut Context<GitBlame>,
     ) {
-        assert_eq!(
+        pretty_assertions::assert_eq!(
             blame
                 .blame_for_rows(
                     &rows
                         .map(|row| RowInfo {
                             buffer_row: Some(row),
+                            buffer_id: Some(buffer_id),
                             ..Default::default()
                         })
                         .collect::<Vec<_>>(),
@@ -697,6 +700,7 @@ mod tests {
             })
             .await
             .unwrap();
+        let buffer_id = buffer.update(cx, |buffer, _| buffer.remote_id());
 
         let git_blame = cx.new(|cx| GitBlame::new(buffer.clone(), project, false, true, cx));
 
@@ -704,12 +708,13 @@ mod tests {
 
         git_blame.update(cx, |blame, cx| {
             // All lines
-            assert_eq!(
+            pretty_assertions::assert_eq!(
                 blame
                     .blame_for_rows(
                         &(0..8)
                             .map(|buffer_row| RowInfo {
                                 buffer_row: Some(buffer_row),
+                                buffer_id: Some(buffer_id),
                                 ..Default::default()
                             })
                             .collect::<Vec<_>>(),
@@ -728,12 +733,13 @@ mod tests {
                 ]
             );
             // Subset of lines
-            assert_eq!(
+            pretty_assertions::assert_eq!(
                 blame
                     .blame_for_rows(
                         &(1..4)
                             .map(|buffer_row| RowInfo {
                                 buffer_row: Some(buffer_row),
+                                buffer_id: Some(buffer_id),
                                 ..Default::default()
                             })
                             .collect::<Vec<_>>(),
@@ -747,12 +753,13 @@ mod tests {
                 ]
             );
             // Subset of lines, with some not displayed
-            assert_eq!(
+            pretty_assertions::assert_eq!(
                 blame
                     .blame_for_rows(
                         &[
                             RowInfo {
                                 buffer_row: Some(1),
+                                buffer_id: Some(buffer_id),
                                 ..Default::default()
                             },
                             Default::default(),
@@ -803,6 +810,7 @@ mod tests {
             })
             .await
             .unwrap();
+        let buffer_id = buffer.update(cx, |buffer, _| buffer.remote_id());
 
         let git_blame = cx.new(|cx| GitBlame::new(buffer.clone(), project, false, true, cx));
 
@@ -813,6 +821,7 @@ mod tests {
             // lines.
             assert_blame_rows(
                 blame,
+                buffer_id,
                 0..4,
                 vec![
                     Some(blame_entry("1b1b1b", 0..4)),
@@ -831,6 +840,7 @@ mod tests {
         git_blame.update(cx, |blame, cx| {
             assert_blame_rows(
                 blame,
+                buffer_id,
                 0..2,
                 vec![None, Some(blame_entry("1b1b1b", 0..4))],
                 cx,
@@ -843,6 +853,7 @@ mod tests {
         git_blame.update(cx, |blame, cx| {
             assert_blame_rows(
                 blame,
+                buffer_id,
                 1..4,
                 vec![
                     None,
@@ -855,7 +866,13 @@ mod tests {
 
         // Before we insert a newline at the end, sanity check:
         git_blame.update(cx, |blame, cx| {
-            assert_blame_rows(blame, 3..4, vec![Some(blame_entry("1b1b1b", 0..4))], cx);
+            assert_blame_rows(
+                blame,
+                buffer_id,
+                3..4,
+                vec![Some(blame_entry("1b1b1b", 0..4))],
+                cx,
+            );
         });
         // Insert a newline at the end
         buffer.update(cx, |buffer, cx| {
@@ -865,6 +882,7 @@ mod tests {
         git_blame.update(cx, |blame, cx| {
             assert_blame_rows(
                 blame,
+                buffer_id,
                 3..5,
                 vec![Some(blame_entry("1b1b1b", 0..4)), None],
                 cx,
@@ -873,7 +891,13 @@ mod tests {
 
         // Before we insert a newline at the start, sanity check:
         git_blame.update(cx, |blame, cx| {
-            assert_blame_rows(blame, 2..3, vec![Some(blame_entry("1b1b1b", 0..4))], cx);
+            assert_blame_rows(
+                blame,
+                buffer_id,
+                2..3,
+                vec![Some(blame_entry("1b1b1b", 0..4))],
+                cx,
+            );
         });
 
         // Usage example
@@ -885,6 +909,7 @@ mod tests {
         git_blame.update(cx, |blame, cx| {
             assert_blame_rows(
                 blame,
+                buffer_id,
                 2..4,
                 vec![None, Some(blame_entry("1b1b1b", 0..4))],
                 cx,

--- a/crates/editor/src/git/blame.rs
+++ b/crates/editor/src/git/blame.rs
@@ -195,9 +195,12 @@ impl GitBlame {
     ) -> impl 'a + Iterator<Item = Option<BlameEntry>> {
         self.sync(cx);
 
+        let buffer_id = self.buffer_snapshot.remote_id();
         let mut cursor = self.entries.cursor::<u32>(&());
         rows.into_iter().map(move |info| {
-            let row = info.buffer_row?;
+            let row = info
+                .buffer_row
+                .filter(|_| info.buffer_id == Some(buffer_id))?;
             cursor.seek_forward(&row, Bias::Right, &());
             cursor.item()?.blame.clone()
         })

